lib/ansible/plugins/strategy/linear.py

```diff
@@ -98,103 +66,40 @@ def _get_next_task_lockstep(self, hosts, iterator):
             host_tasks[host.name] = iterator.get_next_task_for_host(host, peek=True)
         display.debug("done building task lists")
 
-        num_setups = 0
-        num_tasks = 0
-        num_rescue = 0
-        num_always = 0
-
         display.debug("counting tasks in each state of execution")
         host_tasks_to_run = [(host, state_task)
                              for host, state_task in host_tasks.items()
                              if state_task and state_task[1]]
 
         if host_tasks_to_run:
-            try:
-                lowest_cur_block = min(
-                    (iterator.get_active_state(s).cur_block for h, (s, t) in host_tasks_to_run
-                     if s.run_state != IteratingStates.COMPLETE))
-            except ValueError:
-                lowest_cur_block = None
-        else:
-            # empty host_tasks_to_run will just run till the end of the function
-            # without ever touching lowest_cur_block
-            lowest_cur_block = None
-
-        for (k, v) in host_tasks_to_run:
-            (s, t) = v
-
-            s = iterator.get_active_state(s)
-            if s.cur_block > lowest_cur_block:
-                # Not the current block, ignore it
-                continue
-
-            if s.run_state == IteratingStates.SETUP:
-                num_setups += 1
-            elif s.run_state == IteratingStates.TASKS:
-                num_tasks += 1
-            elif s.run_state == IteratingStates.RESCUE:
-                num_rescue += 1
-            elif s.run_state == IteratingStates.ALWAYS:
-                num_always += 1
-        display.debug("done counting tasks in each state of execution:\n\tnum_setups: %s\n\tnum_tasks: %s\n\tnum_rescue: %s\n\tnum_always: %s" % (num_setups,
-                                                                                                                                                  num_tasks,
-                                                                                                                                                  num_rescue,
-                                                                                                                                                  num_always))
-
-        def _advance_selected_hosts(hosts, cur_block, cur_state):
-            '''
-            This helper returns the task for all hosts in the requested
-            state, otherwise they get a noop dummy task. This also advances
-            the state of the host, since the given states are determined
-            while using peek=True.
-            '''
-            # we return the values in the order they were originally
-            # specified in the given hosts array
+            while True:
+                tasks = [task._uuid for host, (state, task) in host_tasks_to_run]
```

Comment:
I like this approach.

You could simplify a bit here by defining tasks outside of the loop instead of `host_tasks_to_run`.

```python
        tasks = [state_task[1]._uuid for state_task in host_tasks.values()
                 if state_task and state_task[1]]

        if tasks:
            while True:
                ...
```

Reply:
👍 took it a bit further and rewrote most of the method :-)